### PR TITLE
Swin transformer

### PR DESCRIPTION
**Pull request by PierreRouge**

### CODE REVIEW

File: collections/_posts/2022-10-03-Swin-Transformer.md [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+* This first block of patch merging is referred as Stage 2
+
+* This process is repeated in Stage 3 ans Stage 4 to obtain a final matrix $$\mathbf{z} \in \mathbb{R}^{\frac{N}{64} \times 8C}$$
+
+  > These steps produce a hierarchical representation similar to a convolutional networks. Therefore Swin Transformers can serve as a general backbone for several vision tasks.
+
+
+# Shifted Window based Self-Attention
+
+* Swin Transformer block is built by replacing the standard multi-head self attention (MSA) module in a Transformer block by a module based on  **shifted windows** with other layers kept the same.
```
Comment (OlivierBernard) [inline collections/_posts/2022-10-03-Swin-Transformer.md:48]: I propose to remove "with other layers kept the same"

File: collections/_posts/2022-10-03-Swin-Transformer.md [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+
+  ![](/collections/images/Swin-Transformer/transformer-block.jpg)
+
+
+
+
+* The self-attention is compute within local windows. The windows are arranged to evenly partition the image in a non-overlapping manner.
+
+* For windows of size M x M patches and an image of h x w patches  the computational complexity of a standard MSA module and a window based self-attention module (W-MSA) are :
```
Comment (OlivierBernard) [inline collections/_posts/2022-10-03-Swin-Transformer.md:57]: patches, the computational

File: collections/_posts/2022-10-03-Swin-Transformer.md [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+> * Swin-B: C = 128, layer numbers ={2, 2, 18, 2}
+> * Swin-L: C = 192, layer numbers ={2, 2, 18, 2}
+
+## Image classification on ImageNet-1k
+
+* Training from scratch on ImageNet-1k 
+  * Swin Transformers surpass others transformers architecture (DeiT architecture and ViT architecture)
+  * And is comparable with convolution based architecture (EfficientNet and Regnet)
+* Pre-training on ImageNet-20k and fine tuning on ImageNet-1k
+  * Pre-training improve the results but it's less significant than ViT
```
Comment (OlivierBernard) [inline collections/_posts/2022-10-03-Swin-Transformer.md:100]: improves

File: collections/_posts/2022-10-03-Swin-Transformer.md [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+
+![](/collections/images/Swin-Transformer/classification-results.jpg) 
+
+## Object Detection on COCO
+
+* Comparison to ResNet and DeiT (transformer) as a backbone for object detection framework
```
Comment (OlivierBernard) [inline collections/_posts/2022-10-03-Swin-Transformer.md:106]: Comparison with

File: collections/_posts/2022-10-03-Swin-Transformer.md [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+
+* First the image of size $$H*W$$ is split in $$N$$ patches of size 4 x 4 to create tokens of size $$4*4*3=48$$
+
+  > ViT have bigger patches of size 16 x 16 ( so less suitable for dense prediction task such as segmentation)
+
+* Then a linear embedding layer is applied to project it to an arbitrary dimensions $$C$$ 
+
+* Two Swin Transfomer blocks (with modified self-attention) are applied giving a new matrix $$\mathbf{z} \in \mathbb{R}^{N \times C}$$
+
+* To produce a hierarchical representation, the number of tokens is reduced by patch merging layers as the network gets deeper
```
[the remark this answers is not in the document]
Comment (PierreRouge) [reply collections/_posts/2022-10-03-Swin-Transformer.md:33]: In the original paper they keep the name "patch merging layer" even if the operation is indeed performed on the token matrix

[the remark this answers is not in the document]
Comment (OlivierBernard) [reply collections/_posts/2022-10-03-Swin-Transformer.md:33]: Ok, so keep it as in the paper :)

File: collections/_posts/2022-10-03-Swin-Transformer.md [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+    title:   "Swin Transformer: Hierarchical Vision Transformer using Shifted Windows"
+    venue:   " IEEE/CVF International Conference on Computer Vision 2021	"
+pdf: "https://arxiv.org/abs/2103.14030"
+---
+
+# Highlights
+
+* ViT has a quadratic complexity with respect to the number of tokens ( and therefore image size). This makes ViT unsuitable for dense prediction task requiring a huge number of tokens such as segmentation.
+* In the proposed architecture the Multi-Head Self Attention block (MSA) is replace by a module based on **shifted windows** (W-MSA) which allow to obtain a **linear complexity** with respect to the number of token. 
+* Also in the ViT architecture the dimension of feature maps is fixed and constant
```
[the remark this answers is not in the document]
Comment (PierreRouge) [reply collections/_posts/2022-10-03-Swin-Transformer.md:17]: In the ViT architecture, the size of the token matrix is constant throughout the network whereas in the Swin architecture the number of token is reduced thanks to the "patch merging layers"
Maybe I can replace this sentence and the following one by:
In the ViT architecture, the size of the token matrix is constant throughout the network, whereas in the proposed architecture the number of token is reduced thanks to the "patch merging layers". This operation allows to build a hierarchical representation.

[the remark this answers is not in the document]
Comment (OlivierBernard) [reply collections/_posts/2022-10-03-Swin-Transformer.md:17]: I let you do the change :)

File: collections/_posts/2022-10-03-Swin-Transformer.md [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+* This process is repeated in Stage 3 ans Stage 4 to obtain a final matrix $$\mathbf{z} \in \mathbb{R}^{\frac{N}{64} \times 8C}$$
+
+  > These steps produce a hierarchical representation similar to a convolutional networks. Therefore Swin Transformers can serve as a general backbone for several vision tasks.
+
+
+# Shifted Window based Self-Attention
+
+* Swin Transformer block is built by replacing the standard multi-head self attention (MSA) module in a Transformer block by a module based on  **shifted windows**
+
+  ![](/collections/images/Swin-Transformer/transformer-block.jpg)
```
Comment (OlivierBernard) [inline collections/_posts/2022-10-03-Swin-Transformer.md:50]: An additional remark: just precise what LN means in the figure (is it layer normalization ?)

**Review by OlivierBernard — CHANGES_REQUESTED** [review]

File: collections/_posts/2022-10-03-Swin-Transformer.md (hunk 7 above)
Comment (PierreRouge) [reply collections/_posts/2022-10-03-Swin-Transformer.md:55]: Yes, it stands for layer normalization, I added the precision in the post ! :)

**Review by OlivierBernard — APPROVED** [review]
Excellent post, thank you Pierre ;)
I will come to see you to have some more details on the architecture, but this is out of the scope of the content of your post, so everything is fine !

**Pull request merged**